### PR TITLE
Add comments and tidy _rewrite function.

### PR DESCRIPTION
**Pull request by odow**
I started fixing a bug for Julia 1.6, and got thoroughly confused
with what this function was doing. Hopefully this is more readable.

### CODE REVIEW

**Comment by blegat** [comment]
Could you split the style fixes and code changes ? It is indeed hardly readable and it's one of the reason I move it to a separate package to at least remove any JuMP-specific stuff and try to make it simpler. I agree that it could still be improved with refactoring.

File: src/rewrite.jl [hunk 1]
```diff
@@ -291,114 +291,279 @@ function _write_add_mul(vectorized, minus, current_sum, left_factors, inner_fact
...
-                    vcat(left_factors, [esc(inner_factor.args[i]) for i in 2:(which_idx - 1)]),
-                    vcat(right_factors, [esc(inner_factor.args[i]) for i in length(inner_factor.args):-1:(which_idx + 1)]),
-                    new_var)
-            else
-                blk = Expr(:block)
+            # A multiplication expression *(args...). We need `!vectorized`
+            # otherwise `x .+ A * b` would be rewritten
+            # `broadcast!(add_mul, x, A, b)`.
+
+            # TODO(odow): remove the special case. I don't know if it's worth
```
Comment (blegat) [inline src/rewrite.jl:386]: Why woudn't it be worth it, I don't feel comfortable doing this change, I would require careful benchmarking and testing with JuMP or a convincing argument :)

Comment (odow) [reply src/rewrite.jl]: Reinstated.

**Comment by odow** [comment]
It's not just style fixes.

**Comment by blegat** [comment]
> It's not just style fixes.

By "not", you mean "now" ?

**Comment by odow** [comment]
> > It's not just style fixes.
> 
> By "not", you mean "now" ?

Haha. Yes.

**Review by blegat — APPROVED** [review]

File: src/rewrite.jl [hunk 2]
```diff
@@ -291,114 +291,285 @@ function _write_add_mul(vectorized, minus, current_sum, left_factors, inner_fact
...
     end
-    # at the lowest level
-    return new_var, _write_add_mul(vectorized, minus, current_sum, left_factors, (esc(inner_factor),), right_factors, new_var)
+    # None of the special cases were hit! This probably means we are vectorized.
+    code = _write_add_mul(
+        vectorized,
+        minus,
+        current_sum,
+        left_factors,
+        (esc(inner_factor),),
```
Comment (odow) [inline src/rewrite.jl:570]: The bug for 1.6 is here. If `inner_factor` contains a `sum(1 for _ = 1:0)`, then this evaluates it and throws an error. I'll think about what extra logic needs to be added to fix it in #60

**Pull request merged**